File: src/main/java/ru/kontur/crypto/money/Envelope.java
package ru.kontur.crypto.money;

import java.math.BigInteger;

public class Envelope {
    private Cheque cheque;

    public Envelope(Cheque cheque){
        this.cheque = cheque;
    }

    public void setSignatureFrom(Bank bank){
        BigInteger k = null;
        while (true){
            long tmp = System.currentTimeMillis();
            if (gcd(tmp, bank.getN()) == 1){
                k = BigInteger.valueOf(tmp);
                break;
            }
        }
        BigInteger m = new BigInteger(cheque.getContent());
        BigInteger e = BigInteger.valueOf(bank.getE());
        BigInteger n = BigInteger.valueOf(bank.getN());

        BigInteger kExpE = k.modPow(e, n);
        BigInteger signedK = (BigInteger.valueOf(bank.sign(kExpE.longValue()))).modInverse(n);

        BigInteger content = (m.multiply(kExpE)).mod(n);
        long result = bank.sign(content.longValue());
        BigInteger signature = (BigInteger.valueOf(result).multiply(signedK)).mod(n);

        cheque.setSignature(signature.longValue());
    }

    private static long gcd (long a, long b){
        return b == 0 ? a : gcd(b, a % b);
    }
}

File: ElectonicMoney/Envelope.cs
using System;
using System.Numerics;

namespace ElectonicMoney
{
    public class Envelope
    {
        public bool Opened { get; private set; }

        private readonly Check check;

        private readonly BigInteger k;

        public Envelope(Check check)
        {
            this.check = check;

            
            while (true)
            {
                var random = new Random(Guid.NewGuid().GetHashCode());
                k = random.Next(2, (int)Bank.Instance.GetN() - 1);
                if (Gcd(k, Bank.Instance.GetN()) == 1)
                {
                    Console.WriteLine($"k = {k}");
                    break;
                }
            }
        }

        private static BigInteger Gcd(BigInteger a, BigInteger b)
        {
            return b == 0 ? a : Gcd(b, a.Mode(b));
        }

        public BigInteger GetContent()
        {
            var res = (check.Content*BigInteger.ModPow(k, Bank.Instance.GetE(), Bank.Instance.GetN())).Mode(
                Bank.Instance.GetN());

            return res;
        }

        public void Sign(BigInteger signature)
        {
            check.Signature = (signature*k.Invert(Bank.Instance.GetN())).Mode(Bank.Instance.GetN());
        }

        public Check Open()
        {
            Opened = true;
            return check;
        }
    }
}